### PR TITLE
tracker(dm): close and recreate tracker when pause and resume

### LINKED ISSUE #5344: after print "rollback checkpoint" in log, unistore folder size grows
### What did you do?

https://github.com/pingcap/tiflow/blob/630896740566eb0f7a81062506e92177e755f404/dm/syncer/checkpoint.go#L715-L727

mock the logic of rollback checkpoint

```
	for i := 0; i < 10000; i++ {
		err = tracker.DropTable(tbl)
		c.Assert(err, IsNil)

		err = tracker.CreateSchemaIfNotExists("test")
		c.Assert(err, IsNil)
		err = tracker.CreateTableIfNotExists(tbl, oriTi)
		c.Assert(err, IsNil)
	}
```

during the test

```
# lance @ LAPTOP-AK9BOS6J in ~/go/src/github.com/pingcap/tiflow on git:630896740
 x [19:38:27]
$ du -h /tmp/tidb-unistore-temp944770732
16M     /tmp/tidb-unistore-temp944770732/kv
16M     /tmp/tidb-unistore-temp944770732

# lance @ LAPTOP-AK9BOS6J in ~/go/src/github.com/pingcap/tiflow on git:630896740
 x [19:38:42]
$ du -h /tmp/tidb-unistore-temp944770732
104M    /tmp/tidb-unistore-temp944770732/kv
104M    /tmp/tidb-unistore-temp944770732

# lance @ LAPTOP-AK9BOS6J in ~/go/src/github.com/pingcap/tiflow on git:630896740
 x [19:40:22]
$ du -h /tmp/tidb-unistore-temp944770732
195M    /tmp/tidb-unistore-temp944770732/kv
195M    /tmp/tidb-unistore-temp944770732
```

So if the task is resumed for many times (which means the task has a resumable error), the size of unistore folder will grow.

### What did you expect to see?

the size of `/tmp/tidb-unistore-temp944770732` has an upper-bound

### What did you see instead?

_No response_

### Versions of the cluster

v5.3.1

### current status of DM cluster (execute `query-status <task-name>` in dmctl)

```console
(paste current status of DM cluster here)
```

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close #5344
fix https://github.com/pingcap/tiflow/issues/3734

### What is changed and how it works?

as title. And when (re)create tracker, load all table structure from checkpoints

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
Fix a bug that after auto resume, DM will use more disk space.
```

### CODE REVIEW

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

**Comment by D3Hunter** [comment]
> To me, this is caused by TableInfo in schema tracker is not consistent with memory table checkpoint. We can fix it when refine the code.

seems this pr can also fix this problem: https://github.com/pingcap/tiflow/issues/5272

**Comment by lance6716** [comment]
```
[2022-05-11T13:11:04.306Z] + echo /tmp/dm_test/s3_dumpling_lighting/worker1/log/stdout.log
[2022-05-11T13:11:04.306Z] /tmp/dm_test/s3_dumpling_lighting/worker1/log/stdout.log
[2022-05-11T13:11:04.306Z] + cat /tmp/dm_test/s3_dumpling_lighting/worker1/log/stdout.log
[2022-05-11T13:11:04.306Z] dm-worker startup 2022-05-11 21:10:47.989560136 +0800 CST m=+0.274499891
[2022-05-11T13:11:04.306Z] panic: success check file not exist!!
[2022-05-11T13:11:04.306Z] 
[2022-05-11T13:11:04.306Z] goroutine 459 [running]:
[2022-05-11T13:11:04.306Z] github.com/pingcap/tiflow/dm/loader.getMydumpMetadata({0x757cd18, 0xc00047ddc0}, 0x14?, 0xc001273b00, {0xc000dbfff6, 0x7})
[2022-05-11T13:11:04.306Z] 	/home/jenkins/agent/workspace/dm_ghpr_integration_test/go/src/github.com/pingcap/tiflow/dm/loader/util.go:132 +0x1007
[2022-05-11T13:11:04.306Z] github.com/pingcap/tiflow/dm/loader.(*LightningLoader).Process(0xc0012b00e0, {0x757cd18, 0xc00047ddc0}, 0x0?)
[2022-05-11T13:11:04.306Z] 	/home/jenkins/agent/workspace/dm_ghpr_integration_test/go/src/github.com/pingcap/tiflow/dm/loader/lightning.go:304 +0x1e5
[2022-05-11T13:11:04.306Z] created by github.com/pingcap/tiflow/dm/dm/worker.(*SubTask).run
[2022-05-11T13:11:04.306Z] 	/home/jenkins/agent/workspace/dm_ghpr_integration_test/go/src/github.com/pingcap/tiflow/dm/dm/worker/subtask.go:260 +0x758
```

File: dm/syncer/schema.go [hunk 1]
```diff
@@ -115,37 +128,21 @@ func (s *Syncer) OperateSchema(ctx context.Context, req *pb.OperateWorkerSchemaR
...
-		if err != nil {
-			return "", terror.ErrSchemaTrackerCannotCreateTable.Delegate(err, sourceTable)
-		}
-
 		s.exprFilterGroup.ResetExprs(sourceTable)
 
-		if !req.Flush && !req.Sync {
-			break
+		if !req.Flush {
+			s.tctx.L().Info("overwrite --flush to true for operate-schema")
```
Comment (GMHDBJD) [inline dm/syncer/schema.go:134]: Now `--flush false` has no effect, we may remove it later(and docs)

File: dm/tests/shardddl4/run.sh [hunk 2]
```diff
@@ -422,6 +422,9 @@ function DM_130_CASE() {
 	run_sql_source2 "insert into ${shardddl1}.${tb1} values(5,5);"
 	run_sql_source2 "insert into ${shardddl1}.${tb2} values(6,6);"
 
+	# make sure 2 DM-workers have a order to see DDL
+	sleep 2
```
Comment (GMHDBJD) [inline dm/tests/shardddl4/run.sh:426]: prefer check log in worker

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Comment by GMHDBJD** [comment]
Also fix #3734

File: dm/pkg/schema/tracker.go [hunk 3]
```diff
@@ -366,6 +372,12 @@ func (tr *Tracker) Reset() error {
...
 func (tr *Tracker) Close() error {
+	if tr == nil {
+		return nil
+	}
+	if !tr.closed.CAS(false, true) {
+		return nil
+	}
 	tr.se.Close()
 	tr.dom.Close()
 	if err := tr.store.Close(); err != nil {
```
Comment (lichunzhu) [inline dm/pkg/schema/tracker.go:383]: Do we need to remove `tr.storePath` before we start? I'm afraid an ungraceful stop may not remove the storage path.

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM

File: dm/pkg/schema/tracker.go (hunk 3 above)
Comment (lance6716) [reply dm/pkg/schema/tracker.go:383]: TempDir will return a newly created folder by appending suffix

Comment (lance6716) [reply dm/pkg/schema/tracker.go:383]: Oh user may need to manually clean it to release disk space. Hope it's not occupy large space.

**Comment by lance6716** [comment]
PTAL @lichunzhu

**Review by lichunzhu — APPROVED** [review]
LGTM

File: dm/pkg/utils/db.go [hunk 4]
```diff
@@ -671,3 +671,10 @@ func GetMaxConnectionsForConn(ctx context.Context, conn *sql.Conn) (int, error)
 func IsMariaDB(version string) bool {
 	return strings.Contains(strings.ToUpper(version), "MARIADB")
 }
+
+// CreateTableSQLToOneRow formats the result of SHOW CREATE TABLE to one row.
+func CreateTableSQLToOneRow(sql string) string {
+	sql = strings.ReplaceAll(sql, "\n", "")
+	sql = strings.ReplaceAll(sql, "  ", " ")
```
Comment (D3Hunter) [inline dm/pkg/utils/db.go:678]: what if there are column defaults contains those char.

seems only used in get schema, ok for now.

**Pull request merged**